control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs

```diff
@@ -130,61 +135,126 @@ async fn nexus_reconciler(
-                    });
-                }
-            }
+        for child in nexus_state.children.iter().filter(|c| c.state.faulted()) {
+            let _ = handle_child_rebuild(nexus, child, &nexus_state, context).await;
         }
-        .instrument(span)
+    }
+    Ok(PollerState::Idle)
+}
+
+/// Checks rebuild policy for the child. Depending on the policy performs the either waits for
+/// child to comeback for specified duration or does Full rebuild straightaway.
+#[tracing::instrument(skip(nexus, context), level = "trace", fields(nexus.uuid = %nexus.uuid, request.reconcile = true))]
+async fn handle_child_rebuild(
+    nexus_spec: &mut OperationGuardArc<NexusSpec>,
+    child: &Child,
+    nexus: &Nexus,
+    context: &PollContext,
+) -> Result<(), SvcError> {
+    let wait_duration = RuleSet::faulted_child_wait(nexus, context.registry());
+    let is_elapsed = is_time_elapsed(child.faulted_at, wait_duration, &child.uri);
+    if child.state_reason == ChildStateReason::RebuildFailed || is_elapsed {
+        info!(%child.uri, "Start full rebuild for child, elapsed: {}, child state reason: {:?}", is_elapsed.to_string(), child.state_reason);
+        faulted_children_remover(nexus_spec, child, context).await?
+    } else if get_child_replica(nexus_spec.as_ref(), child, context)
         .await
+        .is_ok()
+    {
+        info!(%child.uri, "Child's Replica is Online within the partial rebuild window");
+        if let Err(error) = initialize_partial_rebuild(nexus_spec, &child.uri, context).await {
+            info!(%child.uri, "Initializing Full rebuild as online child attempt failed for child, failed with : {:?}", error);
+            faulted_children_remover(nexus_spec, child, context).await?
+        }
+    } else {
+        info!(%child.uri, "Child's Replica is not online yet");
     }
+    Ok(())
+}
 
-    PollResult::Ok(PollerState::Idle)
+/// Removes child from nexus children list to initiate Full rebuild of child.
+#[tracing::instrument(skip(nexus, context), level = "trace", fields(nexus.uuid = %nexus.uuid(), request.reconcile = true))]
+pub(super) async fn faulted_children_remover(
+    nexus: &mut OperationGuardArc<NexusSpec>,
+    child: &Child,
+    context: &PollContext,
+) -> Result<(), SvcError> {
+    let nexus_uuid = nexus.uuid();
+    let nexus_state = context.registry().nexus(nexus_uuid).await?;
+    let nexus_spec_clone = nexus.lock().clone();
+    nexus_spec_clone.warn_span(|| {
+        tracing::warn!(%child.uri, %child.state, %child.state_reason, "Attempting to remove faulted child")
+    });
+    nexus
+        .remove_child_by_uri(context.registry(), &nexus_state, &child.uri, true)
+        .await?;
+
+    nexus_spec_clone.warn_span(|| {
+        tracing::warn!(%child.uri, %child.state, %child.state_reason, "Successfully removed faulted child")
+    });
+    Ok(())
+}
+
+/// Returns true if time elapsed from child fault time is greater or equal to wait time duration.
+fn is_time_elapsed(
```

Comment:
suggest using something like `partial_rebuild_elapsed`

Reply:
I think partial_rebuild_elapsed doesnt indicate wait duration getting elapsed. How about wait_duration_elapsed() ?

Reply:
`wait_duration_elapsed` also doesn't tell you that's for the partial rebuild.

Reply:
We have explained why we wait in the PR comment.